hooks/tk-multi-publish2/basic/collector.py

```diff
@@ -0,0 +1,291 @@
+
+            # store the template on the item for use by publish plugins. we
+            # can't evaluate the fields here because there's no guarantee the
+            # current session path won't change once the item has been created.
+            # the attached publish plugins will need to resolve the fields at
+            # execution time.
+            channel_item.properties["work_template"] = work_template
+            layer_item.properties["work_template"] = work_template
+            self.logger.debug("Work file template defined for Mari collection.")
+
+        return channel_item, layer_item
+
+    def _collect_files(self, settings, channel_item, layer_item):
+        """
+        Collect texture files to be published
+        :param dict settings: Configured settings for this collector
+        :param channel_item:   Parent Item instance for channels
+        :param layer_item:   Parent item instance for layers
+        """
+
+        # check that we are currently inside a project:
+        if not mari.projects.current():
+            raise TankError("You must be in an open Mari project to be able to publish!")
+
+        publisher = self.parent
+
+        icon_path = os.path.join(
+            self.disk_location,
+            os.pardir,
+            "icons",
+            "texture.png"
+        )
+
+        # Look for all layers for all channels on all geometry.  Create items for both
+        # the flattened channel as well as the individual layers
+        for geo in mari.geo.list():
+            geo_name = geo.name()
+            
+            for channel in geo.channelList():
+                channel_name = channel.name()
+
+                # find all publishable layers:
+                publishable_layers = self._find_publishable_layers_r(channel.layerList())
+                if not publishable_layers:
+                    # no layers to publish!
+                    continue
+
+                # add item for whole flattened channel:
+                item_name = "%s, %s" % (geo.name(), channel.name())
+                item = channel_item.create_item(
+                    "mari.texture",
+                    "Channel",
+                    item_name
+                )
+                item.set_icon_from_path(icon_path)
+                item.thumbnail_enabled = True
+                
+                item.properties["geo_publish_name"] = geo_name
+                item.properties["channel_publish_name"] = channel_name
+                item.set_thumbnail_from_path(self._extract_mari_thumbnail())
+
+                # add item for each publishable layer:
+                found_layer_names = set()
+                for layer in publishable_layers:
+                    
+                    # for now, duplicate layer names aren't allowed!
+                    layer_name = layer.name()
+                    if layer_name in found_layer_names:
+                        # we might want to handle this one day...

```

Comment:
does this mean a layer could get overwritten at export time? or possibly publishes with the same name?